compiler/AST/ForallStmt.cpp

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ForallStmt.h"
+#include "AstVisitor.h"
+#include "passes.h"
+
+/////////////////////////////////
+// ForallStmt
+/////////////////////////////////
+
+//
+// This loop:
+//
+//   forall VARS in ITERS with WITH do BODY;
+//
+// is represented by a ForallStmt 'fs' like this:
+//
+// * fs->fMBlock is:
+//
+//   {
+//     def chpl__parIdx      // formerly 'chpl__leadIdx' or 'chpl__saIdx'
+//     def chpl__parIdxCopy
+//     ... helper code to implement forall intents may be here ...
+//     {
+//       // BODY block
+//       ... helper code e.g. to detuple the yielded value may be here ...
+//       ... rest of BODY statements ...
+//     }
+//   }
+//
+// * fs->fIter is ITERS
+//    - a single-element AList for non-zippered loops
+//    - one element per zip component, for zippered loops
+//
+// * fs->fZippered is true for zippered loops
+//
+// * fs->fWith represents the 'with' clause
+//
+
+// convert _build_tuple args to AList
+// todo: have the parser build AList directly or at least keep it as PRIM_ZIP
+// i.e. skip zipToTuple()
+static void convertIterator(AList& list, Expr* iter, bool zippered) {
+  if (zippered) {
+    CallExpr* icall = toCallExpr(iter);
+    INT_ASSERT(icall && icall->isNamed("_build_tuple"));
+    for_actuals(actual, icall)
+      list.insertAtTail(actual->remove());
+  } else {
+    list.insertAtTail(iter);
+  }
+}
+
+ForallStmt::ForallStmt(BlockStmt* var, Expr* iter, bool zippered,
+                       ForallIntents* intents, BlockStmt* body) :
+  Stmt(E_ForallStmt),
+  fMBlock(var),
+  fZippered(zippered),
+  fWith(intents)
+{
+  // Callers must provide these.
+  INT_ASSERT(fMBlock);
+  INT_ASSERT(fMBlock->blockTag == BLOCK_NORMAL);
+  INT_ASSERT(fWith);
+  INT_ASSERT(!body || body->blockTag == BLOCK_NORMAL);
+
+  if (body) {
+    // We are creating a new node from scratch.
+    fMBlock->insertAtTail(body);
+    convertIterator(fIter, iter, zippered);
+  } else {
+    // Creating a new node when most contents already exist,
+    // ex. when copying an existing node.
+    INT_ASSERT(!iter);
```

Comment:
I'd generally prefer to have a different ForallStmt constructor for copying vs creating one during parsing.